Move devxp linux to ~/.local/devxp-files, and fix permission issue

Kardzhilov opened the pull request:
## Issues

- The original issue was to move the directory to the `~/.local/devxp-files` location to align with the structure of our other projects.
- It was discovered that there are permission issues of the last version of `devxp-linux` break `git pull`, preventing us from pushing out an update that will fix it in the future.

## Proposed Solution
Initially, I attempted to automate the moving of the directory. However, during testing, I encountered the permission issues mentioned above. Strangely, these permission issues were not displayed as errors, and I only discovered them because changes I made to the repository were not being picked up during updates.

Fixing this issue to enable the script to perform a `git pull` would require the user to execute multiple commands to take ownership of the directory and mark it as a `safe.directory`. However, since this would essentially involve setting up the directory just to remove it afterward, I believe it's more practical for users to simply clone the repository again.

## Changes Made
To address these issues, I have added a step in the process that checks for the existence of `/usr/local/bin/devxp-files` and cleans it up, including any references to it in `.bashrc` or other initialization files. Additionally, I have modified the script to use the `~/.local/devxp-files` going forward.

michaeloa commented on README.md:
```diff
@@ -43,7 +43,7 @@ wsl.exe --install -d Ubuntu
 
 #### Step 1: Clone Repository
-The `devxp-linux` repository will automatically clone itself under `/usr/local/bin/devxp-files` once executed, so where you initially clone it is not crucial. Execute the following command to clone the repository:
+The `devxp-linux` repository will automatically clone itself under `$HOME/.local/devxp-files` once executed, so where you initially clone it is not crucial. Execute the following command to clone the repository:
```
Let's make this $HOME/.local/devxp/

michaeloa commented on README.md:
```diff
@@ -57,7 +57,7 @@ To install necessary developer tools and receive weekly update reminders, execut
 The script will ask for your **sudo password** and then your **BECOME password**. Those are both the same password, which is your Ubuntu WSL password.
 
-> The reminder frequency can be changed by modifying the `cooldown` value in the `/usr/local/bin/devxp-files/cooldown.sh` script **after** the auto-setup is complete. By default it is set to 7 days.
+> The reminder frequency can be changed by modifying the `cooldown` value in the `$HOME/.local/devxp-files/cooldown.sh` script **after** the auto-setup is complete. By default it is set to 7 days.
```
Should be in .local/devxp/bin or .local/devxp/scripts

But incidentally, wouldn't it make sense to use a .cooldown data file, like you do for timestamp, et al?

Alternatively, we can create a .devxp config file that people can edit. We're going to need one for the gh cli extension anyway, so possible reuse the same?

Kardzhilov replied:
- Yes we could put it in a bin or scripts folder.
- The reason why i did not make it a dot file is that i wanted it to be easier to find for a potential user.
- This cooldown script gets called from the rc file so it runs on each terminal startup, i doubt we would want that for the gh cli file?

michaeloa commented on auto-setup.sh:
```diff
@@ -34,18 +34,37 @@ else
+old_path="/usr/local/bin/devxp-files"
+new_path="$HOME/.local/devxp-files"
+
+# Check if the current directory is the old directory
```
Comment that this can be removed after some grace period. Maybe create a debt ticket as well.

michaeloa commented on scripts/auto-bootstrap.sh:
```diff
@@ -6,11 +6,13 @@ RED=$(tput setaf 1)
 YELLOW=$(tput setaf 3)
 NC=$(tput sgr0) # No Color
-timestamp="/usr/local/bin/devxp-files/.timestamp"
+timestamp="$HOME/.local/devxp-files/.timestamp"
```
I would place this files in their own place.
.local/devxp/data/.timestamp ?

Then user would be .local/devxp/data/.user ?

michaeloa commented on scripts/clone.sh:
I think this should have a more  explicit name. devxp-clone, devxp-update, maybe?

michaeloa commented on scripts/cooldown-do-not-edit.sh:
I don't think you need to state this in the name.

Kardzhilov replied:
The reasoning is to avoid confusion where you think you have updated the cooldown value but then you have not.
I originally did not have it as its own file, it was originally formatted as a long string in a file that wrote the file to the location it should be. In a past code review it came up that it was messy so i scrapped that and went with this solution instead.

michaeloa commented on scripts/delete-old-directory.sh:
```diff
@@ -0,0 +1,73 @@
+
+# Elvish
+if command -v elvish &> /dev/null; then
+    remove_line "$HOME/.elvish/rc.elv" "$script_to_remove"
```
Fun. Haven't run into this before.

Is anyone using it? :D

Kardzhilov replied:
I don't know to be honest, i just looked up the 10 or so most popular shell programs haha

michaeloa commented:
This is excellent, and way overdue that we create a devxp in .local. After this, we can move the lint configs in there as well.

michaeloa commented on scripts/update.sh:
This can be run manually as well, right? In that case, it might also go directly into .local/bin.

Kardzhilov replied:
If you want to run it manually you would most likely trigger that using `bootstrap.sh` or `scripts/auto-bootstrap.sh` 
Moving or copying either outside of the repo and into .local/bin would mean that we would need to copy/move them each time we make an update. 
If we want to run an update manually I would instead suggest that we add an alias along the lines of `devxp-update` that triggers `$HOME/.local/devxp/devxp-linux/scripts/auto-bootstrap.sh`. That way it also updates the timestamp at the end.

michaeloa reviewed (approved):
We should probably shift some of the scripts into ansible, but that can be done in a future update.

Kardzhilov commented:
- Moved all the relavant .files into `~/.local/devxp/data`
- Created a simlink from `~/.local/devxp/devxp-linux/scripts/auto-bootstrap.sh` to `~/.local/devxp/scripts/auto-bootstrap.sh`
- Moved the ansible log file to `~/.local/devxp/log` and renamed it `devxp-linux-ansible-playbook.log`
- Created a common `config.yml` file that can be used across projects. For this project it only looks for a `cooldown` variable

Kardzhilov commented on scripts/devxp-clone.sh:
```diff
@@ -0,0 +1,20 @@
+USER=$(cat $HOME/.local/devxp/data/.user)
+
+echo "${GREEN}Cloning repository to $HOME/.local/devxp...${NC}"
+git clone https://github.com/elhub/devxp-linux.git $HOME/.local/devxp/devxp-linux
```
This needs to be changed to `git clone -b move https://github.com/elhub/devxp-linux.git $HOME/.local/devxp/devxp-linux` if local testing is performed. Otherwise you will not pull this branch

sachin170622 commented on scripts/root.sh:
```diff
@@ -16,7 +16,7 @@ fi
 
 echo "${GREEN}Cloning repository to /usr/local/bin/devxp-files...${NC}"
-git clone https://github.com/elhub/devxp-linux.git /usr/local/bin/devxp-files/devxp-linux
+git clone -b move https://github.com/elhub/devxp-linux.git /usr/local/bin/devxp-files/devxp-linux
```
Can we use git clone --mirror here?
